### DISCUSSION TITLE
Pico RTC Alarm?

### DISCUSSION THREAD

**Discussion by gwynge** [micropython/micropython#9594] [General]
Documentation here [https://docs.micropython.org/en/latest/rp2/quickref.html#real-time-clock-rtc](url) suggests that the RP2 has a RTC and refers to machine.RTC documentation. I see that RTC works in terms of setting/getting date time.

However allegedly it includes an "alarm" function. Except on the Pico when you try and use it it says "AttributeError: 'RTC' object has no attribute 'alarm'"

So if it hasn't got an alarm what's the process for performing long running tasks on a schedule? When I say long running I still mean subsecond. But too long to use a timer (especially when I have a 20mS timer going off - a long running process on a second timer stops that happening).

rp2-pico-w-20221011-unstable-v1.19.1-532-g8e912a501

**Comment by scruss** [comment]
While the RP2040's RTC does have an an alarm function (see RP2040 datasheet "4.8.5.4. Configuring an Alarm", p.553), it's not implemented in MicroPython for the RP2. The generic documentation for [machine.RTC()](https://docs.micropython.org/en/v1.19.1/library/machine.RTC.html#machine-rtc) is a superset of what each port provides

**Reply by gwynge** [reply]
Thanks for confirming; do you have a suggestion for how one schedules slightly longer running activities? In this instance I want to run something every 30 minutes that will save collected data to a file. In the meantime I am monitoring something every 20mS which cannot be interrupted.

I would expect to be able to use a 20mS timer for the monitoring and then a scheduled task for the data storage. And the 20mS timer must be able to interrupt the other activity of course.

**Comment by jahr** [comment]
Hello, as the RTC Alarm is not available for Pico and current [API](http://docs.micropython.org/en/latest/library/machine.RTC.html) is not optimal, I've written a simple wrapper around the RP Pico SDK functions for the RTC Alarm. You may check it at https://github.com/jahr/rp2-rtc-alarm.git

**Reply by gwynge** [reply]
That looks interesting, do you know how it interacts with Timers? It's unclear from the doc I've found (or at least read) so far how RTC alarm, timers and Irqs interact and take priority etc. Do you have any knowledge on the matter?

**Reply by jahr** [reply]
I use it for running a clock. So the RTC alarm fires every minute and I'm simultanously using timer firing every 1 s. So far I've seen no problems. IMHO as far as you use `micropython.schedule` and ensure that the code execution ends before another one starts, there should be no problem. 
But if you need to run your code every 20 ms and another every 30 minutes without fixing it to concrete time (which would be rather difficult for the alarm, which easily fires eg. every minute, hour, day etc., but every 30 minutes would require re-seting it, at least in my wrapper ;-) ) I would recommend to use 2 Timer instances instead. With `micropython.schedule` it should not interfere.
